Crud book

juandbc opened the pull request:
## Summary
* Implemeted CRUD methods to books resource
* Fix book find by author feature
* Fix variable references in book model

## Trello Card
https://trello.com/c/5SJmFAv7/18-crear-las-operaciones-crud-de-book

Commit:
Fix variable references in Book setters

Commit:
Add book exceptions
Add exception handler
Add CRUD methods to book controller
Fix book find by author method

Gianni-Nicolas commented on src/main/java/wolox/training/controllers/BookController.java:
```diff
@@ -1,18 +1,70 @@
+    private final BookRepository bookRepository;
+
+    @Autowired
+    public BookController(BookRepository bookRepository) {
```
This constructor is not required. With adding the @Autowired annotation on the attribute
`private BookRepository bookRepository` is sufficient.

juandbc replied:
I think it is better to use constructor injection, because we can know what the dependencies of a class are. As for field injection, it is more difficult to know the dependencies of a class. In addition, field injection makes it difficult to perform unit tests because of all the complexity of reflection.

Gianni-Nicolas replied:
ok great, as a comment I add that the tests can be done in the same way, if you inject the dependency in the attribute. Spring has tools like Mockito and others to perform your unit tests. This will be seen later in the training.

Gianni-Nicolas commented on src/main/java/wolox/training/exceptions/BookIdMismatchException.java:
```diff
@@ -0,0 +1,20 @@
+        super("The book's id does not correspond to the data to update");
+    }
+
+    public BookIdMismatchException(final String message, final Throwable cause) {
```
I don't think this constructor is necessary, leaving the constructor that accepts a custom message as a parameter is enough.

Gianni-Nicolas commented on src/main/java/wolox/training/exceptions/BookIdMismatchException.java:
```diff
@@ -0,0 +1,20 @@
+        super(message);
+    }
+
+    public BookIdMismatchException(final Throwable cause) {
```
I don't think this constructor is necessary, leaving the constructor that accepts a custom message as a parameter is enough.

Gianni-Nicolas commented on src/main/java/wolox/training/exceptions/BookNotFoundException.java:
```diff
@@ -0,0 +1,18 @@
+
+    public BookNotFoundException() {}
+
+    public BookNotFoundException(final String message, final Throwable cause) {
```
I don't think this constructor is necessary, leaving the constructor that accepts a custom message as a parameter is enough.

juandbc replied:
I agree, but I need to implement it becasuse I'm using method reference in `Book b = bookRepository.findById(id).orElseThrow(BookNotFoundException::new);`

Gianni-Nicolas commented on src/main/java/wolox/training/exceptions/BookNotFoundException.java:
```diff
@@ -0,0 +1,18 @@
+        super(message);
+    }
+
+    public BookNotFoundException(final Throwable cause) {
```
I don't think this constructor is necessary, leaving the constructor that accepts a custom message as a parameter is enough.

juandbc replied:
I agree, but I need to implement it becasuse I'm using method reference in `Book b = bookRepository.findById(id).orElseThrow(BookNotFoundException::new);`

Gianni-Nicolas replied:
To use an exception like that the empty constructor should be sufficient.

Another way to call an exception is

bookRepository.findById(id).orElseThrow(() -> new BookNotFoundException());

juandbc replied:
You're right. I'll refactor it

Gianni-Nicolas commented on src/main/java/wolox/training/repositories/BookRepository.java:
```diff
@@ -1,11 +1,9 @@
 public interface BookRepository extends JpaRepository<Book, Long> {
 
-    Book findByAuthor(String author);
+    Book findOneByAuthor(String author);
```
Add that the output of this repository is Optional to avoid doing null checks in the future. (Optional<Book>)

I leave a link to have some additional information on the subject:

- [Optionals](https://www.baeldung.com/java-optional)

Gianni-Nicolas reviewed (changes requested):
Very good job I leave some simple comments to solve!!!

Commit:
Modify return type for findOneByAuthor
Remove unnecessary constructors for BookIdMismatchException
Apply the above changes in BookController

Commit:
Refactor use of BookNotFoundException
